### ISSUE TITLE
Add Docker images as an Ansible location

### ISSUE THREAD

**Issue by mdklatt** [mdklatt/idea-ansible-plugin#10]
Add support for executing Ansible from within a Docker container, *i.e.* `docker run --rm`. This will require a new [project-level settings](https://github.com/mdklatt/idea-ansible-plugin/issues/9) group to specify the Docker image name and optionally the path to the Docker executable.

Add a new `CommandLine` extension method to convert the current instance to a `docker run` command. In this case, the Ansible location will be relative to the Docker container. Environment variables will be passed along to the container so that a virtualenv installation will work. The configuration working directory (should be a project setting?) will be mounted to the container as `/tmp/ansible-<something unique>` (could be a project Docker setting?), which will become the container working directory. This could be problematic if Ansible needs to install dependencies in the local working directory; in general, IDEA solves this by running toolchain containers as `root` so that they can write persistent build artifacts to a local directory.

**Comment by mdklatt** [comment]
Commit 5767110c5371486f4dd3a59b992827cf664446f1 contains what should be a working implementation of this, *i.e.* it generates the correct `docker run` command. It still needs some testing with an actual Docker image that can run ansible.

**Comment by mdklatt** [comment]
With anecdotal testing, the current implementation does not work with a Python virtualenv. The Docker image has Ansible 6.1.0 is installed into a Python 3.10 virtualenv. The error message about permission issues with the temporary directory seems to be a red herring, because the actual problem is specifying `PATH` using `docker --env`. This seemingly has nothing to do with shell substitution, because `--env PATH='/opt/ansible/bin:$PATH'` and `--env PATH='/opt/ansible/bin` produce the same error.


```
$ docker run --rm --env PATH='/opt/ansible/bin:$PATH' --env PYTHONHOME='' 
--env ANSIBLE_CONFIG='/tmp/ansible/ansible.cfg' --env VIRTUAL_ENV='/opt/ansible' 
--workdir /tmp/ansible --volume $PWD:/tmp/ansible --entrypoint ansible-playbook tmp/ansible:6.1.0 
--inventory /tmp/ansible/hosts.yml /tmp/ansible/playbook.yml

fatal: [localhost]: UNREACHABLE! => {"changed": false, "msg": "Failed to create temporary directory. 
In some cases, youmay have been able to authenticate and did not have permissions on the target 
directory. Consider changing the remote tmp path in ansible.cfg to a path rooted in \"/tmp\", for more 
error information use -vvv. Failed command was: ( umask 77 && mkdir -p \"` 
echo /root/.ansible/tmp `\"&& mkdir \"` echo /root/.ansible/tmp/ansible-
tmp-1713996446.3167603-11-239056324019027 `\" && echo ansible-
tmp-1713996446.3167603-11-239056324019027=\"` echo /root/.ansible/tmp/ansible-
tmp-1713996446.3167603-11-239056324019027 `\" ), exited with result 127", "unreachable": true}
```

This is the equivalent command without specifying `$PATH`, and this is successful:

```
$ docker run --rm --env PYTHONHOME='' --env ANSIBLE_CONFIG='/tmp/ansible/ansible.cfg' 
--env VIRTUAL_ENV='/opt/ansible' --workdir /tmp/ansible --volume ${PWD}:/tmp/ansible 
--entrypoint /opt/ansible/bin/ansible-playbook tmp/ansible:6.1.0 --inventory 
/tmp/ansible/hosts.yml /tmp/ansible/playbook.yml


PLAY [Test play for idea-ansible-plugin] ***************************************

TASK [Gathering Facts] *********************************************************
ok: [localhost]

TASK [Contact the host] ********************************************************
ok: [localhost]

PLAY RECAP *********************************************************************
localhost                  : ok=2    changed=0    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   

```

This an alternate version of the command that works where environment variables are embedded in the the container command instead of using `docker --env`. This is a similar approach to [idea-remotepython-plugin](https://github.com/mdklatt/idea-remotepython-plugin).

```
$ docker run --rm  --workdir /tmp/ansible --volume ${PWD}:/tmp/ansible tmp/ansible:6.1.0 sh 
-c  'PATH=/opt/ansible/bin:$PATH VIRTUAL_ENV=/opt/ansible PYTHONHOME="" ANSIBLE_CONFIG=/tmp/ansible/
ansible.cfg ansible-playbook --inventory /tmp/ansible/hosts.yml /tmp/ansible/playbook.yml'
```

**Comment by mdklatt** [comment]
As of commit 4f177a08be7ef409bc307ad7460ca22c28b8a0e2, Docker execution is fully functional (including virtualenvs) for both Playbook and Galaxy run configurations.

There is still some remaining trickiness regarding dependency installation in a Docker container via Galaxy. The basic assumption is that input files (playbooks, inventories, requirement files) are mounted from the local machine to the container, while installed roles and collections can be internal to the container. Ideally, dependencies are built into the image itself rather than installed at runtime. The plugin will mount the local working directory (`workDir`) to the container, so anything within that directory can be accessed by Ansible. If the roles and collections directories are relative to `workDir`, dependencies will be persistent becasue the are writting to the host (but the container requires write permission to the host in this case). If the roles and collections directories are specified with an absolute path, this must be a valid *container* path. In this case, the installed dependencies will not be saved to the host, and must be installed every time the container is run.

For best results, dependencies should be installed into the image when it's built, and then configured via environment variable or *ansible.cfg*. Do not rely on these directories being in the container working directory, because for now this is an implementation detail and not user-configurable.

**Comment by mdklatt** [comment]
Commit cae922d45d6b055837b70beed9dd45a3886f2d9f is an updated README that includes documentation for Docker execution as well as plugin settings and creating run configurations from context menus. 

What else needs to be done for this issue or release? There are still some rough edges around Docker execution, particularly for Galaxy configurations, but some operational experience is going to be required to figure out the best way to do things.

**Comment by mdklatt** [comment]
Issue resolved by commit 88ac402e449d87c1681df5259c494d3799b40025.
